### PR TITLE
coprocessor/dag/expr: add some un-pushed builtin UDFs(LTrim, RTrim)

### PR DESCRIPTION
**Pull request by spongedu**
## What have you changed? (mandatory)

implement builtin function: `LTrim`, `RTrim`  in `coprocessor`.
This is for #3275

## What are the type of the changes? (mandatory)

Improvement

## How has this PR been tested? (mandatory)

unittest

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 1]
```diff
@@ -26,6 +27,26 @@ impl ScalarFunc {
...
     }
+
+    pub fn ltrim<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let val = try_opt!(self.children[0].eval_string(ctx, row));
+        let s = String::from_utf8(val.into_owned())?;
+        Ok(Some(Cow::Owned(s.trim_left_matches(' ').as_bytes().to_vec())))
```
[the remark this answers is not in the document]
Comment (siddontang) [reply src/coprocessor/dag/expr/builtin_string.rs]: better to define a const, but I am fine with this too.

**Comment by spongedu** [comment]
TiDB deals with `\r\n\t` differently with MySQL on `LTrim`, `RTrim` and `Trim`.  I've fix this in [tidb:#7291](https://github.com/pingcap/tidb/pull/7291)

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: ok, I'll define a constant to replace `' '` here.

**Comment by AndreMouche** [comment]
Seems you need run `make format` before commit

**Comment by breeswish** [comment]
Thanks a lot!

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 2]
```diff
@@ -37,6 +39,31 @@ impl ScalarFunc {
...
         Ok(Some(Cow::Owned(format!("{:b}", i).into_bytes())))
     }
+
+    #[inline]
+    pub fn ltrim<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let val = try_opt!(self.children[0].eval_string(ctx, row));
```
Comment (hicqu) [inline src/coprocessor/dag/expr/builtin_string.rs]: Call `eval_string_and_decode` is better than combine `eval_string` and `String::from_utf8`.

Comment (hicqu) [reply src/coprocessor/dag/expr/builtin_string.rs]: I have tested it with MySQL. Seems we don't need parse raw bytes to string here. Just trim the byte slice is OK.
```sql
create table x (a binary(32));
insert into x values (x'ffff');
select hex(trim(a)) from x
+------------------------------------------------------------------+
| hex(trim(a))                                                     |
+------------------------------------------------------------------+
| FFFF000000000000000000000000000000000000000000000000000000000000 |
+------------------------------------------------------------------+
1 row in set (0.00 sec)
```

Comment (hicqu) [reply src/coprocessor/dag/expr/builtin_string.rs]: Same behavior with `TiDB`.

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: @hicqu  I've  refined the implementation here. PTAL :)

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 3]
```diff
@@ -37,6 +39,35 @@ impl ScalarFunc {
...
+        let pos = val.iter().position(|&x| x != SPACE);
+        if let Some(i) = pos {
+            Ok(Some(Cow::Owned(val[i..].to_vec())))
```
Comment (hicqu) [inline src/coprocessor/dag/expr/builtin_string.rs]: Here you can write
```rust
            match val {
                Cow::Borrowed(val) => Ok(Some(Cow::Borrowed(&val[i..]))),
                Cow::Owned(val) => Ok(Some(Cow::Owned(val[i..].to_owned()))),
            }
```
to reduce some mem alloc.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 4]
```diff
@@ -37,6 +39,35 @@ impl ScalarFunc {
...
+        } else {
+            Ok(Some(val))
```
Comment (hicqu) [inline src/coprocessor/dag/expr/builtin_string.rs]: I guess it should be `Ok(Some(Cow::Owned(b"".to_vec())))`.

**Comment by hicqu** [comment]
rest LGTM. Thanks a lot. :)

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 3 above)
Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: OK, thx, I'll refine here

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 4 above)
Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: Yes, this is a mistake. I've fixed here and add some more test cases

**Comment by spongedu** [comment]
@hicqu PTAL :)

**Comment by hicqu** [comment]
LGTM.

**Comment by zz-jason** [comment]
the calculation logic LGTM.

**Comment by huachaohuang** [comment]
@spongedu Please resolve conflicts.

**Comment by spongedu** [comment]
@huachaohuang conflicts resolved. PTAL :)

**Pull request merged**